### ISSUE TITLE
BigGAN: targeted data augmentation of hard features (figures/hands)

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#29]
See #19. Targeted data augmentation by cropping images down to specific difficult features can help sculpt GAN outputs by making such images more common, and providing much larger versions which devote many more pixels (many of which may not have been seen at a mere 512px resolution) with fewer distractions. This worked well with faces.

There are two major areas where our anime BigGAN samples fall down: overall figure coherency (there are still difficulties with number of limbs, geometry, consistent shapes) and especially hands. Hands were a problem with the TWDNE portraits as well - hands are common enough that GANs must try (and fail) to generate them, but also something humans are highly attuned to (so low quality is unacceptable), yet also small highly-varied regions of the image (so hard to learn no matter what).

I have gotten [AniSeg to crop images by 'figures'](https://github.com/jerryli27/AniSeg/issues/6) ([samples](https://imgur.com/a/khLvVsD)) and am creating a cropped subset of Danbooru SFW now. (It is sufficiently slow that I'm not sure it'd be feasible to incorporate any segmenter into the TPU pod IO pipeline: delivering 4096 segmented images per iteration using a large slow CNN model is not an easy task, and we may be much better off just creating specialized datasets as needed, because it'll be faster and far simpler to implement.)

For hands, there are no anime off-the-shelf croppers, and the existing human/photographic hand detectors will probably fail. We can use Shawn's web GUI to train YOLO like how Arfa did faces and other body parts for e621 for TFDNE, and we can filter down using the Danbooru tag [`hands`](https://danbooru.donmai.us/posts?tags=hands+rating%3As) (*n* ~ 19k) to spend annotating time on the larger higher-quality hands available. Probably 1-2k annotations will be enough (we don't need super-high accuracy because any erroneous crop will still be a crop of a real image). I can start work on that as soon as the web GUI is set up.

Open question: after faces/figures/hands, what else would be worth making annotations for to crop out & data augment?

**Comment by gwern** [comment]
Update: we got `hands` annotation working last night and did ~1k ([JSON export](https://www.dropbox.com/s/37mlu647znko7ax/2020-05-15-tensorfork-experiments-boundingboxes-hands.jsonl?dl=0)). The tag works well and should yield good training data because of the wide variety of poses and hand sizes and styles. 

Unfortunately, Shawn's rating tool currently has a design bug where it shows the same images in the same sequence to all users in a given 'experiment', so we cannot crowdsource ratings, as this would merely lead to people annotating the same few dozen or hundred images repeatedly instead of collectively annotating thousands of images, and redundant bounding box annotations are (almost completely) uninformative compared to a new image. So we are blocked on either fixing that or a single person doing them all.

**Comment by shawwn** [comment]
It's @sheminusminus's rating tool. Also, it's updated now to shuffle the images, which over any reasonable-sized dataset will reduce the duplicated work to < 20%.

A simulation:

![image](https://user-images.githubusercontent.com/59632/82132784-4fed5b00-9798-11ea-916f-2a2a4acfe56c.png)

**Comment by gwern** [comment]
After another evening and full day of labeling, plus a few hundred from others, we now have 9248 annotated hands: https://www.dropbox.com/s/717qs2puvgzz2h8/2020-05-17-tensorfork-experiments-boundingboxes-hands.jsonl?dl=0

**Comment by gwern** [comment]
Hand status: currently blocked on training YOLO. We now have _n_=14,252 hands annotated.

To generate a larger dataset of pony faces beyond what he could crop out with a furry face detector from e621 (to improve his StyleGAN model and create 'This Pony Does Not Exist' as a followup to [TFDNE](https://thisfursonadoesnotexist.com/)), Arfa has begun working on a pony face detector in the same vein aimed at cropping Derpibooru faces, with the bounding box annotations being done at: https://www.tagpls.com/exp?n=derpibooru-s

**Comment by gwern** [comment]
Figures writeup: https://www.gwern.net/Crops#danbooru2019-figures

**Comment by gwern** [comment]
Another intriguing possibility besides ranker (#30 ) for data cleaning/augmentation: using anime style transfer to generate multiple differently-colored variants of each image. 

["ImageNet-trained CNNs are biased towards texture; increasing shape bias improves accuracy and robustness"](https://arxiv.org/abs/1811.12231), Geirhos et al 2018 ([repo](https://github.com/rgeirhos/Stylized-ImageNet)):

> Convolutional Neural Networks (CNNs) are commonly thought to recognise objects by learning increasingly complex representations of object shapes. Some recent studies suggest a more important role of image textures. We here put these conflicting hypotheses to a quantitative test by evaluating CNNs and human observers on images with a texture-shape cue conflict. We show that ImageNet-trained CNNs are strongly biased towards recognising textures rather than shapes, which is in stark contrast to human behavioural evidence and reveals fundamentally different classification strategies. We then demonstrate that the same standard architecture (ResNet-50) that learns a texture-based representation on ImageNet is able to learn a shape-based representation instead when trained on "Stylized-ImageNet", a stylized version of ImageNet. This provides a much better fit for human behavioural performance in our well-controlled psychophysical lab setting (nine experiments totalling 48,560 psychophysical trials across 97 observers) and comes with a number of unexpected emergent benefits such as improved object detection performance and previously unseen robustness towards a wide range of image distortions, highlighting advantages of a shape-based representation.

As noted about StyleGAN's Laplacian pyramid, StyleGAN seems to excel on anime in part by learning to draw lines and *then* filling them in with color. BigGAN, for its other virtues, does not appear to do this and tends to take a more conventionally-CNN-like approach of learning textures/colors and sharpening edges later. If we could do style transfer to generate several versions of Danbooru2019 with different colors for each image by picking a random image as the 'style' source and generating a style-transferred version (we would want to do this offline as style transfer is usually quite slow compared to regular data augmentations like cropping), it might make BigGAN learn much better and more varied images.

However, the only good classifier or tagger for anime images right now is [DeepDanbooru](https://github.com/KichangKim/DeepDanbooru), which no one has used for style transfer yet. DeepDanbooru is also a fairly small and minimally trained NN (I believe it is a Resnet-50 trained for just 1 GPU-month on the top 10k tags or something like that). It is nowhere near the [SOTA](https://paperswithcode.com/task/image-classification) for classification, which would be something like [EfficientNet](https://github.com/tensorflow/tpu/tree/master/models/official/efficientnet#2-using-pretrained-efficientnet-checkpoints) or [FixRes](https://github.com/facebookresearch/FixRes). With our Danbooru2019 dataset & TPU pods, we could probably train a *much* better model by transfer learning from public SOTAs.

**Comment by gwern** [comment]
Update: I have trained and released a YOLOv3 model, and have adapted Arfa's script to hand cropping Danbooru2019. I have ~65k hands cropped out so far.

**Comment by gwern** [comment]
[*PALM*: the PALM Anime Locator Model](https://www.gwern.net/Crops#hands) is now officially released. ~58k hands should go a long way to fixing the uncanny valley of bad hands in BigGAN samples!

With both [Figures](https://www.gwern.net/Crops#danbooru2019-figures) and PALM now released, this issue is largely resolved unless another obvious category of data augmentable objects pops up, so I will close it.
